# ĐỒ ÁN CUỐI KỲ MÔN LẬP TRÌNH CHO KHOA HỌC DỮ LIỆU

*Giảng viên: Thầy Trần Trung Kiên*

---

**Nhóm sinh viên thực hiện:**

Sinh viên 01: Trương Trọng Lộc - 18120197

Sinh viên 02: Dương Nam Hải - 19120068

---

## Import các thư viện cần thiết

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

---

## Thu thập dữ liệu

- Tập dữ liệu sử dụng trong bài được lấy từ nguồn link (https://www.kaggle.com/mmattson/who-national-life-expectancy, phân tích về các yếu tố liên quan cố thể gây ảnh hưởng đến mức tuổi thọ trung bình của 1 quốc gia, một châu lục trong 1 năm nhất định nào đó.
- Tập dữ liệu được đăng tải bởi người dùng MMattson, với license từ Creative Common Public Domain - CC0, có thể tùy ý sử dụng mà không cần được cấp phép.
- Tập dữ liệu được thu thập từ 2 nguồn là GHO - Global Health Observatory và UNESCO, khảo sát 183 quốc gia trên khắp các khu vực và trong suốt giai đoạn từ năm 2000 đến 2016.
- Tác giả đã thực hiện thu thập thủ công cho phần dữ liệu về UNESCO từ trang web UNESCO Databases of Resources on Education.
- Tác giả đã dùng thư viện request thông qua GHO Data API portal để thu thập dữ liệu từ GHO.
- Cụ thể về các bước thu thập dữ liệu: https://www.kaggle.com/mmattson/template-to-generate-the-dataset

---

## Khám phá dữ liệu

### Đọc dữ liệu từ file csv vào dataframe

In [2]:
data = pd.read_csv("Data/who_life_exp.csv")


### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?

In [3]:
num_rows, num_cols = data.shape
num_rows, num_cols

(3111, 32)

=> Vậy dữ liệu có 3111 dòng và 32 cột.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng trong dữ liệu là thông tin dự đoán tuổi thọ trung bình qua từng năm của một quốc gia dựa vào các cột dữ liệu thể hiện nhiều yếu tố khác nhau.

### Dữ liệu có các dòng bị lặp không?

In [4]:
if data.duplicated().sum() != 0:
    print("Có dòng dữ liệu bị lặp")
else:
    print("Không có dữ liệu trùng lặp")


Không có dữ liệu trùng lặp


### Mỗi cột có ý nghĩa gì?

|Tên cột dữ liệu | Mô tả |
|:------:| ---|
|Country| Tên quốc gia khảo sát|
|country_code| Mã rút gọn của quốc gia, gồm 3 chữ cái|
|region| Lục địa của đất nước thuộc vào|
|year| Năm khảo sát|
|life_expect| Số năm tuổi dự đoán khi vừa sinh|
|life_exp60| Số năm tuổi còn lại dự đoán khi đạt 60 tuổi|
|adult_mortality|Tỷ lệ tử vong ở người trưởng thành (xác suất tử vong từ 15 đến 60 tuổi trên 1000 dân số)|
|infant_mort|Tỷ lệ tử vong dưới 1 tuổi|
|age1-4mort|Tỷ lệ tử vong từ 1 đến 4 tuổi|
|alcohol|Mức tiêu thụ rượu bình quân đầu người (15+)|
|bmi|BMI trung bình (ước tính chuẩn hóa theo độ tuổi)|
|age5-19thinness|Tỷ lệ gầy trẻ em và thanh thiếu niên gầy|
|age5-19obesity	|Tỷ lệ trẻ em và thanh thiếu niên béo phì|
|hepatitis|Tỷ lệ tiêm chủng viêm gan B (HepB) ở trẻ 1 tuổi (%)|
|measles|Tỷ lệ tiêm chủng liều đầu tiên (MCV1) có chứa vắc xin sởi cho trẻ 1 tuổi (%)|
|polio| Tỷ lệ tiêm chủng bại liệt (Pol3) ở trẻ 1 tuổi (%)|
|diphtheria|Tỷ lệ tiêm chủng bạch hầu, uốn ván và ho gà (DTP3) cho trẻ 1 tuổi (%)|
|basic_water|Tỷ lệ dân số được tiếp cận dịch vụ nước sạch tối thiểu|
|doctors|Bác sĩ y khoa (trên 10.000)|
|hospitals|Số lượng bệnh viện/100.000 dân|
|gni_capita|Thu nhập quốc dân trên đầu người (tổng thu thập mà người dân và doanh nghiệp của quốc gia kiếm được)|
|gghe-d	|Chi tiêu từ ngân sách chính phủ cho lĩnh vực y tế tính theo tỷ lệ phần trăm của tổng sản phẩm quốc nội (GDP) (%)|
|che_gdp|Chi phí người dân bỏ ra cho dịch vụ y tế (CHE) tính theo phần trăm tổng sản phẩm quốc nội (GDP) (%)|
|une_pop|Dân số (hàng nghìn)|
|une_infant|Tỷ lệ tử vong, trẻ sơ sinh (trên 1.000 trẻ sống)|
|une_life|Tuổi thọ trung bình|
|une_hiv|Tỷ lệ nhiễm HIV, tổng số (\% dân số từ 15-49 tuổi)|
|une_gni|Thu nhập quốc dân trên đầu người|
|une_poverty|Tỷ lệ số người nghèo có mức sống 1,90 đô la một ngày (PPP) (\% dân số)|
|une_edu_spend|Chi tiêu của chính phủ cho giáo dục tính theo phần trăm GDP (\%)|
|une_literacy|Tỷ lệ người lớn biết chữ, dân số trên 15 tuổi, cả hai giới (\%)|
|une_school|Số năm đi học trung bình (ISCED 1 trở lên), dân số từ 25 tuổi trở lên, cả hai giới|

**Các thuộc tính có tên `une_`** được thống kê bởi tổ chức UNESCO.

**Các thuộc tính còn lại** được thống kê bởi GHO (Global Health Observatory).


### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [5]:
data.dtypes

country             object
country_code        object
region              object
year                 int64
life_expect        float64
life_exp60         float64
adult_mortality    float64
infant_mort        float64
age1-4mort         float64
alcohol            float64
bmi                float64
age5-19thinness    float64
age5-19obesity     float64
hepatitis          float64
measles            float64
polio              float64
diphtheria         float64
basic_water        float64
doctors            float64
hospitals          float64
gni_capita         float64
gghe-d             float64
che_gdp            float64
une_pop            float64
une_infant         float64
une_life           float64
une_hiv            float64
une_gni            float64
une_poverty        float64
une_edu_spend      float64
une_literacy       float64
une_school         float64
dtype: object

Ta sẽ tiến hành xem dữ liệu thật sự của những cột là object xem chúng có kiểu dữ liệu gì

In [6]:
def open_object_dtype(s):
    dtypes = set(s.apply(lambda x: type(x)).unique().tolist())
    return dtypes

In [7]:
object_cols = data.select_dtypes(include = 'object').columns
datatype = object_cols.map(lambda x: open_object_dtype(data[x]))
pd.Series(datatype, index = object_cols)

country         {<class 'str'>}
country_code    {<class 'str'>}
region          {<class 'str'>}
dtype: object

Có lẽ như các cột dữ liệu không có cột nào có kiểu dữ liệu chưa hợp lý cần phải chuyển đổi

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?

In [8]:
def missing_ratio(series):
    return series.isna().mean() * 100

num_df = data.select_dtypes(exclude='object')
pd.concat([num_df.agg([missing_ratio]), num_df.describe()])

year  life_expect   life_exp60  adult_mortality  \
missing_ratio     0.000000     0.000000     0.000000         0.000000   
count          3111.000000  3111.000000  3111.000000      3111.000000   
mean           2008.000000    69.146384    18.912060       193.527786   
std               4.899767     9.129761     2.885412       114.007447   
min            2000.000000    36.227360    10.728000        49.201850   
25%            2004.000000    63.200095    16.619485       108.269750   
50%            2008.000000    71.597200    18.508260       164.822200   
75%            2012.000000    75.537030    21.099175       250.832300   
max            2016.000000    84.166160    26.394020       696.874200   

               infant_mort   age1-4mort      alcohol          bmi  \
missing_ratio     0.000000     0.000000     1.607200     1.092896   
count          3111.000000  3111.000000  3061.000000  3077.000000   
mean              0.032496     0.003489     4.834781    25.052714   
std               0.030361     0.005146     3.998741     2.193445   
min               0.001470     0.000065     0.000000    19.800000   
25%               0.008255     0.000355     1.198030    23.300000   
50%               0.019995     0.000895     3.993560    25.500000   
75%               0.051720     0.004877     7.723420    26.500000   
max               0.164515     0.039095    20.182460    32.200000   

               age5-19thinness  age5-19obesity  ...      che_gdp  \
missing_ratio         1.092896        1.092896  ...     3.760849   
count              3077.000000     3077.000000  ...  2994.000000   
mean                  5.312447        5.972278  ...     6.110353   
std                   4.652408        4.530812  ...     2.505267   
min                   0.100000        0.100000  ...     1.025160   
25%                   1.800000        2.000000  ...     4.238798   
50%                   3.800000        5.200000  ...     5.758030   
75%                   7.800000        8.900000  ...     7.850327   
max                  28.100000       26.700000  ...    20.413410   

                    une_pop   une_infant     une_life      une_hiv  \
missing_ratio  1.189328e+00     0.000000     0.000000    23.818708   
count          3.074000e+03  3111.000000  3111.000000  2370.000000   
mean           3.707550e+04    30.488428    68.957352     2.038397   
std            1.378377e+05    27.756930     9.359194     4.593804   
min            7.601600e+01     1.600000    39.441000     0.100000   
25%            2.195105e+03     8.000000    62.844500     0.100000   
50%            8.544297e+03    19.500000    71.409000     0.400000   
75%            2.509552e+04    48.050000    75.570645     1.500000   
max            1.414049e+06   142.400000    83.984880    28.200000   

                     une_gni  une_poverty  une_edu_spend  une_literacy  \
missing_ratio       3.760849    70.652523      41.337191     81.645773   
count            2994.000000   913.000000    1825.000000    571.000000   
mean            14964.832999    10.845893       4.532930     81.984472   
std             17495.137508    17.168798       1.754280     19.665588   
min               420.000000     0.100000       0.787440     14.376040   
25%              2970.000000     0.600000       3.262830     72.701130   
50%              8340.000000     3.100000       4.425410     90.953740   
75%             20482.500000    12.400000       5.494980     95.786975   
max            122670.000000    94.100000      14.059080     99.998190   

               une_school  
missing_ratio   74.124076  
count          805.000000  
mean             9.712214  
std              2.854723  
min              0.559340  
25%              7.735940  
50%             10.270400  
75%             12.070610  
max             14.378800  

[9 rows x 29 columns]

### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?

In [9]:
def num_diff_vals(series):
    return series.nunique()
def value_ratios(series):
    return dict((series.value_counts()))

cate_df = data.select_dtypes(include = 'object')
cate_col_profiles_df = cate_df.agg([missing_ratio, num_diff_vals, value_ratios])
cate_col_profiles_df

country  \
missing_ratio                                                0.0   
num_diff_vals                                                183   
value_ratios   {'Angola': 17, 'Netherlands': 17, 'Poland': 17...   

                                                    country_code  \
missing_ratio                                                0.0   
num_diff_vals                                                183   
value_ratios   {'AGO': 17, 'NLD': 17, 'POL': 17, 'PRT': 17, '...   

                                                          region  
missing_ratio                                                0.0  
num_diff_vals                                                  6  
value_ratios   {'Europe': 850, 'Africa': 799, 'Americas': 561...

Sau khi khám phá sự phân bố của các cột dữ liệu, ta nhận thấy có 4 cột numeric với missing ratio trên 50% là "hospitals", "une_poverty", "une_literacy", "une_school". Việc dùng các cột nào trong phân tích dữ liệu có thể cho ra kết quả không tốt, vì vậy ta cần loại bỏ chúng trong bước tiền xử lý.

### Tiền xử lý dữ liệu ###

In [10]:
missing_df = data.agg(missing_ratio)
cols_to_drop = missing_df[missing_df > 50]
cols_to_drop

hospitals       95.821279
une_poverty     70.652523
une_literacy    81.645773
une_school      74.124076
dtype: float64

In [11]:
data_df = data.drop(cols_to_drop.index.tolist(), axis = 1)
data_df.shape

(3111, 28)

---

## Đưa ra câu hỏi cần trả lời

### Tiền xử lý

### Phân tích dữ liệu để trả lời cho các câu hỏi 

---

## Nhìn lại quá trình làm đồ án

---

## Tài liệu tham khảo

---